Coordinator throttling RFC (#42)
prestodb/rfcs: RFC-0011-coordinator-throttling.md

For immediate relief, we propose implementing a holistic admission control mechanism that provides reactive throttling of the overall workload by not admitting queries.

### Goals
- Decrease duration of overload on the cluster
- Also decrease overall variance of execution time of queries (queuing is fine)

## Proposed Implementation
### Worker side changes
- Added endpoint to gather worker load (`/v1/info/nodestats`)
  - Include NodeState + worker load. Idea is to just call `/v1/info/nodestats` rather than `/v1/info/state` (eventually deprecate it). Worker load is already getting populated in [cpp workers](https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/main/PrestoServer.cpp#L1537).  
- PR: https://github.com/prestodb/presto/pull/25686

### Coordinator side changes
#### Collecting worker load data
  - Update DiscoveryManager to invoke this end point rather than the `/v1/info/state`
  - Expose `getNodeLoadMetrics` on InternalNodeManager
  - PR: https://github.com/prestodb/presto/pull/25688 
#### Scheduling policies
  - Implement node overload policies that can govern the policies to determine cluster overload

>   - Based on these policies and load collected from worker, improve the admission control logic globally to not queue the query (irrespective of Resource Group)

Reviewer: Some folks rely on the JMX connector to monitor metrics.  I think there should be flexibility on how to define these limits so that we can separate them out with a resource group configuration (for example, it's a common resource group configuration to separate out DDL statements from other queries, precisely because DDLs don't use worker resources).

A good example is `ResourceGroupQueryLimits`.  This is a per-query limit that is defined at the RG level.  I could see this throttling specification being configured similarly.

>   - Based on these policies and load collected from worker, improve the admission control logic globally to not queue the query (irrespective of Resource Group)

Reviewer: DDLs and system queries too.  For example, if I need to query the information_schema table, it should not queue at all--none of that is going to the workers, it's just metadata.

**Few Examples**:
- **Intermediate stage**: For 333 hash partitions, randomly select 333 workers from the top 500 scoring workers. (We run with 300 node clusters so this property needs to be tuned as well if we go with this approach.)
- **Leaf stage**: For leaf task, the logic in `chooseLeastBusyNode` could be augmented to pick workers with best score and out of these assign the splits.

- **Pros**
This could provide load balancing to begin with and prevent some workers getting hot

- **Cons**
In our cluster most of the workers become overloaded in a fraction of minutes due to a lot of queries getting admitted. 

Load balancing may not be immediately helpful in such cases
This could be a good direction for the future if we see only a select few workers are causing issues.

### Approach 2
Write a separate service which can consume worker load and update global RG (max queries that can run based on the load / policies)

- **Pros**
No coordinator changes

- **Cons**
Extra service to maintain
Two brains

## Adoption Plan

> If this feature is enabled, this can cause queuing when cluster is loaded (which would be separate from queuing due to Resource Groups)

Reviewer 2: ```suggestion
Enabling this feature can cause additional queuing when the cluster is overloaded (which would be separate from queuing due to Resource Groups).
```

> . For now end users can correlate queuing via looking at queuing + worker overload metrics together. If required, we can also expose separate queuing metrics due to cluster overload.

Author: Sounds good. I will add these metrics. Updated the section as well.

## Test Plan

> Added the test cases

Reviewer 2: Spell this out more about what kinds of tests you are adding.

Author: Added the details on testing.

This strategy does not consider the current system load when assigning intermediate tasks. 

#### Resource Group throttling
There are certain metrics like softMemoryLimit etc which exist at RG level to queue the workload, but it does not consider the actual utilization and is out of scope for this discussion.

While we could improve task scheduling (pro-active measure to not let worker overload) through load balancing (more on this in alternative approaches), this alone would be insufficient given the nature of our overload scenario where all workers are saturated in short period of time which is due to admitting lot of heavy queries in short duration of time. 

For immediate relief, we propose implementing a holistic admission control mechanism that provides reactive throttling of the overall workload by not admitting queries.

### Goals
- Decrease duration of overload on the cluster
- Also decrease overall variance of execution time of queries (queuing is fine)

## Proposed Implementation
### Worker side changes

> - Added endpoint to gather worker load (`/v1/info/nodestats`)

Reviewer: Can you include what is returned from this endpoint?

Reviewer 3: +1

It would be good to document the fields (string constants) here or in the "## Metrics" section and refer to it.

Author: Added the response and field

  - Include NodeState + worker load. Idea is to just call `/v1/info/nodestats` rather than `/v1/info/state` (eventually deprecate it). Worker load is already getting populated in [cpp workers](https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/main/PrestoServer.cpp#L1537).  
- PR: https://github.com/prestodb/presto/pull/25686

### Coordinator side changes
#### Collecting worker load data
  - Update DiscoveryManager to invoke this end point rather than the `/v1/info/state`
  - Expose `getNodeLoadMetrics` on InternalNodeManager
  - PR: https://github.com/prestodb/presto/pull/25688 
#### Scheduling policies
  - Implement node overload policies that can govern the policies to determine cluster overload

>   - Based on these policies and load collected from worker, improve the admission control logic globally to queue the query (irrespective of Resource Group)

Reviewer: There isn't a whole lot of detail here.  Please be more specific--how is overload determined?

Reviewer 3: @tdcmeehan 
Each worker is determining if it is overloaded according to the code and the config parameters.
Coordinator is merely notifier is the worker has memory or CPU overload at the time of reporting.

The worker code to determine the overload can now be thought of being version 1.
We think to improve it as it is not 'snappy' enough.

We could create some documentation and link to it here.
Would need advice where we can create a doc page describing worker overload detection and pushback and how verbose it should be.

Author: Added some details, but as @spershin  mentioned the majority of load logic is in cpp worker. I added link to committed PRs, but happy to update doc / create one.

Reviewer: Rather than a binary overloaded/not-overloaded, have you considered reporting raw metrics and allowing a user to self-determine what is a sensible overloaded value (with a sensible default)?

I am thinking that it would first make sense to queue on traditional resource group metrics, as these can have finer grained control and can isolate problematic users.  You would use this as a backstop in case that were not effective.  But they should be defined in conjunction with the resource group metrics, as the overload decision should involve metrics that exceed whatever was configured in the resource group.  For example, memory limits only rely on what is reported--for unreported memory, we use this as a backstop.

By the way, what sort of unreported memory is being allocated in Velox?

Author: > I am thinking that it would first make sense to queue on traditional resource group metrics, as these can have finer grained control and can isolate problematic users.

Do you mean tune current RG configs to not let cluster overload to begin with?

> defined in conjunction with the resource group metrics, as the overload decision should involve metrics that exceed whatever was configured in the resource group. 
Not sure if I understood this part. Can you provide more details?

> By the way, what sort of unreported memory is being allocated in Velox?

Velox has different mechanism than java for memory allocation. I think RSS + cached memory is unaccounted for but @spershin may provide more details.

Reviewer 3: @tdcmeehan 

Thanks for the comments!

We have some (I would call them) blockers for making this framework tied to RGs at the moment.
I'll try to describe them here.

1. In the worker we do not have a good understanding of how much memory each query uses due to the untraceable memory (this cannot be fixed now).
The cpu usage snapshot (not total so far) would be also hard if not impossible to collect per query, although we could measure something like how many active execution threads each query has. This might be good to consider in the long run.

2. The idea of not allowing queries to be started from the coordinator is an extension of the worker pushback framework. The latter is just each worker postpones starting task if the worker is overloaded. This creates certain ambiguities for query metrics: this postponing time contributes to the execution time rather than queued time, the different queries might be queued in different order on different workers, etc.
To help with it we notify coordinator about the overload, so we can have a central entity queuing the majority of the queries during the overload times.
So, it is not so much as for the coordinator to manage it at the moment, although in the future we might do something like that.

3. RGs are not fair at all - they do not really consider CPU and memory most of the time: memory logic is there, but since we cannot report proper memory from the worker it is not really reflecting the situation. In other words it is mainly based on the number of queries and we know that all queries aren't equal.

4. Queuing ALL queries can be considered fair if we don't look at priorities.

We propose to add the suggestion to the RFC as the potential to improve the framework and continue with the planned implementation for the phase 1.

As for the other and more detailed metrics - it is possible. We have map<string, double> in the endpoint structure and it can have anything we want in the future. For the phase 1 we thought to keep it simple.

Author: @tdcmeehan  - I have added sections for

1. Using raw metrics
2. Improvement as extension of this RFC (In Approach-1)
3. Query configs to throttle at query level

Let me know if this looks good to you

  - PR: https://github.com/prestodb/presto/pull/25689 

## Metrics

> - Cluster overload

Reviewer: Is this a binary metric?  Can you elaborate?

Author: Added details on the JMX metrics added.

> - Variance in runtime of queries

Reviewer: Where and how is this calculated?

![Design diagram](RFC-0011/cpuload.png)

The impact is:

- Large variance in execution time
- Memory kills / Overload on dependency services like metastore / warm-storage etc 

This provide more reasons to work on protecting cluster overload.

## Background

While not exhaustive, current throttling/possible mechanisms relevant for this discussion

#### Granular task based scheduling

**Leaf Task scheduling**

The node selection logic in Presto involves randomly selecting 10 nodes from those that do not currently have scheduled splits for the given query. Among these 10 nodes, the system then chooses the node that is the least busy. For reference, the relevant code implementation can be found [here](https://github.com/prestodb/presto/blob/master/presto-main-base/src/main/java/com/facebook/presto/execution/scheduler/nodeSelection/SimpleNodeSelector.java#L182).
Additionally, in the process of leaf task scheduling, a feedback mechanism is employed whereby splits are assigned to the nodes deemed least busy. It is important to note that in this context, "busyness" is defined by the number of splits being processed by a node, rather than its actual system load. 

> While this approach is generally effective for distributing work, it may not always result in the most optimal allocation of resources.

Reviewer: Can you elaborate on the difference between "effectiveness when distributing work" vs. "allocating resources"?  To me they sound the same--resources are worker resources, so effective distribution of work means effective allocation of resources.

Author: In our scenario even with this logic of choosing least busy nodes based on splitting, most of the workers could become overload. I didn't meant to say this logic is wrong. Rather over-admitting is the key issue here.

For immediate relief, we propose implementing a holistic admission control mechanism that provides reactive throttling of the overall workload by not admitting queries.

### Goals
- Decrease duration of overload on the cluster
- Also decrease overall variance of execution time of queries (queuing is fine)

## Proposed Implementation
### Worker side changes
- Added endpoint to gather worker load (`/v1/info/nodestats`)
  - Include NodeState + worker load. Idea is to just call `/v1/info/nodestats` rather than `/v1/info/state` (eventually deprecate it). Worker load is already getting populated in [cpp workers](https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/main/PrestoServer.cpp#L1537).
  - **Fields**
    - **nodeState** [Code link](https://github.com/prestodb/presto/blob/master/presto-spi/src/main/java/com/facebook/presto/spi/NodeState.java)  
    - **metrics**
      - Map<String, Double> loadMetrics
        - key is overload state like `cpu.overload`, `memory.overload`

>         - val is 0 or 1 to indicate overload or not

Author: Yes, will make it an object as updated in the newer version

While this approach is generally effective for distributing work, it may not always result in the most optimal allocation of resources.

**Intermediate task scheduling**

For intermediate task scheduling in Presto, the selection of workers is based on a hash partition count. Specifically, intermediate tasks are assigned to worker nodes according to the hash of the partition, which determines the distribution across available intermediate workers. The relevant code implementation can be found [here](https://github.com/prestodb/presto/blob/master/presto-main-base/src/main/java/com/facebook/presto/sql/planner/SystemPartitioningHandle.java#L164).
This strategy does not consider the current system load when assigning intermediate tasks. 

#### Resource Group throttling
There are certain metrics like softMemoryLimit etc which exist at RG level to queue the workload, but it does not consider the actual utilization and is out of scope for this discussion.

> While we could improve task scheduling (pro-active measure to not let worker overload) through load balancing (more on this in alternative approaches), this alone would be insufficient given the nature of our overload scenario where all workers are saturated in short period of time which is due to admitting lot of heavy queries in short duration of time. 

Author: I was referring to the task scheduling. I have updated the description to make it accurate.

For immediate relief, we propose implementing a holistic admission control mechanism that provides reactive throttling of the overall workload by not admitting queries.

### Goals
- Decrease duration of overload on the cluster
- Also decrease overall variance of execution time of queries (queuing is fine)

## Proposed Implementation
### Worker side changes
- Added endpoint to gather worker load (`/v1/info/nodestats`)

>   - Include NodeState + worker load. Idea is to just call `/v1/info/nodestats` rather than `/v1/info/state` (eventually deprecate it). Worker load is already getting populated in [cpp workers](https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/main/PrestoServer.cpp#L1537).

Reviewer: Can this plan include the deprecation?  I think we need to clean up the old endpoint as well so we're not confusingly returning two different things depending on whether or not we're running a C++ cluster.

Author: Added the deprecation plan

- Variance in runtime of queries
  - **Cluster Overload Patterns:** We have identified clusters experiencing consistent overload patterns that cause significant variation in query runtime performance.
  - **Measurement Strategy:** We will measure end-to-end execution time improvements after implementing query admission controls to reduce the overall query load using this feature.

## Other Approaches Considered

### Approach 1
Improve the resource group accounting to include metrics that actually cause overload and improve admission control
Current limitations
- CPU is not considered correctly for throttling
- Memory is not accounted correctly on Velox

Given the limitations, the work here would be to be report the metrics that actually matter and attribute as accurately as it could be
For e.g. for cpu, we may need to look for a proxy metrics like `numQueuedDrivers` etc.

- **Pros**
  - RG throttling code could be reused
  - More fair to throttle workloads which are actually overloading

- **Cons**
  - Needs significant work on worker side
  - Also need work on coordinator side
  - Need more analysis on key metrics

>   - No enforcement: Even if we fix the accounting and improve admission control to run the cluster less loaded, there is no guarantee that RGs won't exceed their limits, potentially causing some workers to become overloaded since Presto lacks preemption support and queries can use more resources once admitted.

Author: Yes you are right - we are not enforcing in this model either. I only brought this up in context of RG accounting fairness guarantees. I am ok framing current RFC as second line of support mechanism. Do you recommend I remove RG from alternative approach section and update the `Background` section with these details

>   - No enforcement: Even if we fix the accounting and improve admission control to run the cluster less loaded, there is no guarantee that RGs won't exceed their limits, potentially causing some workers to become overloaded since Presto lacks preemption support and queries can use more resources once admitted.

Author: Updated the RFC per this comment.
